keylime/cloud_verifier_tornado.py

```diff
@@ -150,31 +222,50 @@ def delete(self):
 
         if agent_id is None:
             common.echo_json_response(self, 400, "uri not supported")
-            logger.warning('DELETE returning 400 response. uri not supported: ' + self.request.path)
-
-        agent = self.db.get_agent(agent_id)
+            logger.warning(
+                'DELETE returning 400 response. uri not supported: ' + self.request.path)
+        try:
+            agent = session.query(VerfierMain).filter_by(agent_id=agent_id).first()
+        except SQLAlchemyError as e:
+                logger.error(f'SQLAlchemy Error: {e}')
 
         if agent is None:
             common.echo_json_response(self, 404, "agent id not found")
-            logger.info('DELETE returning 404 response. agent id: ' + agent_id + ' not found.')
+            logger.info('DELETE returning 404 response. agent id: ' +
+                        agent_id + ' not found.')
             return
 
-        op_state =  agent['operational_state']
+        op_state = agent.operational_state
         if op_state == cloud_verifier_common.CloudAgent_Operational_State.SAVED or \
-        op_state == cloud_verifier_common.CloudAgent_Operational_State.FAILED or \
-        op_state == cloud_verifier_common.CloudAgent_Operational_State.TERMINATED or \
-        op_state == cloud_verifier_common.CloudAgent_Operational_State.TENANT_FAILED or \
-        op_state == cloud_verifier_common.CloudAgent_Operational_State.INVALID_QUOTE:
-            self.db.remove_agent(agent_id)
+                op_state == cloud_verifier_common.CloudAgent_Operational_State.FAILED or \
+                op_state == cloud_verifier_common.CloudAgent_Operational_State.TERMINATED or \
+                op_state == cloud_verifier_common.CloudAgent_Operational_State.TENANT_FAILED or \
+                op_state == cloud_verifier_common.CloudAgent_Operational_State.INVALID_QUOTE:
+            try:
+                session.query(VerfierMain).filter_by(agent_id=agent_id).delete()
+                session.commit()
+            except SQLAlchemyError as e:
+                logger.error(f'SQLAlchemy Error: {e}')
             common.echo_json_response(self, 200, "Success")
-            logger.info('DELETE returning 200 response for agent id: ' + agent_id)
+            logger.info(
+                'DELETE returning 200 response for agent id: ' + agent_id)
         else:
-            self.db.update_agent(agent_id, 'operational_state',cloud_verifier_common.CloudAgent_Operational_State.TERMINATED)
-            common.echo_json_response(self, 202, "Accepted")
-            logger.info('DELETE returning 202 response for agent id: ' + agent_id)
-
+            try:
+                update_agent = session.query(VerfierMain).get(agent_id)
+                update_agent.operational_state = cloud_verifier_common.CloudAgent_Operational_State.TERMINATED
+                try:
+                    session.add(update_agent)
```

Comment:
What does the `add` method here do?  It seems like we're trying to update the operational state of an existing entry, so do we need to be doing some sort of update instead, or can `add` be overloaded this way?

Reply:
Think of this as an update. Alchemy uses add for also updating row information. So we first of all  ORM Query object on line 254 and then set the operational state to failed for that object as `TERMINATED`  then reload the now updated object `update_agent` into the session.

There is an (albeit old) explanation [here](https://groups.google.com/forum/#!topic/sqlalchemy/ukM3UbHI5KA) and another [here](https://overiq.com/sqlalchemy-101/crud-using-sqlalchemy-orm/#updating-data)

Reply:
Ok, just making sure this wasn't a typo ... I'm still getting used to SQLAlchemy and how it works.

Reply:
> Ok, just making sure this wasn't a typo ... I'm still getting used to SQLAlchemy and how it works.

Me too and the documents are not strongly opinionated. I hope in time others will refactor if there are better ways of doing things. can't see any performance impact at all from when I decorated with a timing function.